### DISCUSSION TITLE
How to set server side timeout?

### DISCUSSION THREAD

**Discussion by claudydetry** [jchristn/SuperSimpleTcp#89] [Q&A]
Hello Joel,

Thank you for developing SimpleTcp.
We are using it as a server, but I have a question.

It seems to disconnect around six seconds after receiving a keep alive message from the client which send it every 10 seconds.
How can we modify the server timeout?

Best regards.
Claudy

**Comment by jchristn** [comment]
Hi @claudydetry thanks for your note.  Have you set ```server.IdleClientTimeoutMs``` to the appropriate value? 

For example:
```
server.IdleClientTimeoutMs = 30000;
```
Will set it to disconnect clients after 30 seconds.

**Comment by claudydetry** [comment]
Hi @jchristn,

I just set the server.IdleClientTimeoutMs to 30000 now but the behavior remains the same.

The 'keep alive' messages generate a ClientConnected event but rarely a DataReceived event.
Once a DataReceived event is triggered, the server remains connected and the following 'keep alive' messages generate DataReceived events.

Sometimes, we receive the first message in a few seconds and sometimes it can take hours.
Once the first message is received, all works fine.

Would there be a way to force the connection?

**Reply by jchristn** [reply]
Hmmm..  Can you reproduce this behavior with the test projects in the source code?  Alternatively if you want to post code to reproduce, I can look at it

**Reply by claudydetry** [reply]
Hello @jchristn,

I could reproduce a similar behavior running your test project on a PC.

Sometimes, when the Keep-Alive message is sent, the server triggers the connection event, then the disconnection event without receiving the message. Sometimes, the message is received properly.
See SimpleTcp-Log file:
[SimpleTcp-Log.txt](https://github.com/jchristn/SimpleTcp/files/7039452/SimpleTcp-Log.txt)

Using another server solution, all the messages are received.
See LCDemo-Log file:
[LCDemo-Log.txt](https://github.com/jchristn/SimpleTcp/files/7039460/LCDemo-Log.txt)

**Reply by jchristn** [reply]
I'll need to see code in order to debug it.  Thanks

**Reply by claudydetry** [reply]
Dear Joel, 

The code is your code. The only thing I changed is the IP address.
The Keep-Alive messages are coming from a third party hardware device, so we have no control at all on that.
Those messages are received without any loss by third party solutions (see LCDemo-Log.txt above).

I suspect that the problem is coming from the transmission timing.
What else can be different between two identical transmission?
How can a message trigger a ClientConnected event but not a DataReceived event?

SimpleTcp is very close to meet our needs and we would be very happy to sponsor it.

**Reply by jchristn** [reply]
Is the client device expecting an HTTP response?  Is your server code sending a 200/OK with the data as shown in the ```LCDemo-Log.txt``` file?
```
<DATA_DL time="10:09:19.821" remoteEndPoint="192.168.51.81:42436" status="OK">
    HTTP/1.1 200 OK
    Content-Type: text/xml; charset=ISO-8859-1
    Content-Length: 61

    <?xml version="1.0" encoding="utf-8"?>
    <stream id="10070" />
```
If the code is indeed my server test code, I can see why this is failing, because you need to be sending the appropriate response back to the requestor.

**Reply by jchristn** [reply]
Also which app is generating the ```LCDemo-Log.txt``` file?

**Reply by claudydetry** [reply]
Yes, the client device expects the 200/Ok with the data shown but that is not critical in the case of the Keep-Alive messages.
My server code is sending the 200/Ok as required but is struggling to get the first message.
The App generating the LCDemo-Log.txt file has been sent by email.

**Reply by jchristn** [reply]
Ok, when I receive the email I'll take a look.
